doc/api/n-api.md

```diff
@@ -3705,6 +3730,93 @@ NAPI_EXTERN napi_status napi_get_uv_event_loop(napi_env env,
 - `[in] env`: The environment that the API is invoked under.
 - `[out] loop`: The current libuv loop instance.
 
+<!-- it's very convenient to have all the anchors indexed -->
+<!--lint disable no-unused-definitions remark-lint-->
+## Asynchronous Thread-safe Function Calls
+JavaScript functions can normally only be called from a native addon's main
+thread. If an addon creates additional threads then N-API functions that require
+a `napi_env`, `napi_value`, or `napi_ref` must not be called from those threads.
+
+This API provides the type `napi_threadsafe_function` as well as APIs to create,
+destroy, and call objects of this type. `napi_threadsafe_function` creates a
+permanent reference to a `napi_value` that holds a JavaScript function, and
+uses `uv_async_t` from libuv to coordinate calls to the JavaScript function from
+all threads.
+
+Since `uv_async_t` is used in the implementation, the caveat whereby multiple
+invocations on the secondary thread may result in only one invocation of the
+JavaScript function also applies to `napi_threadsafe_function`.
+
+### napi_create_threadsafe_function
+<!-- YAML
+added: REPLACEME
+-->
+```C
+NAPI_EXTERN napi_status
+napi_create_threadsafe_function(napi_env env,
+                                napi_value func,
+                                void* data,
+                                size_t argc,
+                                napi_threadsafe_function_marshal marshal_cb,
+                                napi_threadsafe_function_process_result
+                                    process_result_cb,
+                                napi_threadsafe_function* result);
+```
+
+- `[in] env`: The environment that the API is invoked under.
+- `[in] func`: The JavaScript function to call from another thread.
+- `[in] data`: Optional data to attach to the resulting `napi_threadsafe_function`.
+- `[in] context`: Optional context associated with `data`.
+- `[in] argc`: Number of arguments the JavaScript function will have.
+- `[in] marshal_cb`: Optional callback to convert `data` and `context` to
+JavaScript function arguments. The callback will always be called on the main
+thread.
+- `[in] process_result_cb`: Optional callback to handle the return value and/or
+exception resulting from the invocation of the JavaScript function. The callback
+will always be called on the main thread.
+- `[out] result`: The asynchronous thread-safe JavaScript function.
+
+### napi_call_threadsafe_function
+<!-- YAML
+added: REPLACEME
+-->
+```C
+NAPI_EXTERN napi_status napi_call_threadsafe_function(napi_threadsafe_function func);
+```
+
+- `[in] func`: The asynchronous thread-safe JavaScript function to invoke. This
+API may be called from any thread.
+
+### napi_get_threadsafe_function_data
```

Comment:
This function and `napi_delete_threadsafe_function()` are poorly named, IMO.  The "threadsafe" suggests they're, well, thread-safe, but they're not - no locking, no synchronization, nothing.

Reply:
Is it not sufficient that the `void*` is established before any thread is started, and is read-only after that? Perhaps I need to add this recommendation to the documentation.

As for the delete, I guess my intention was that it happen after all threads using the `napi_threadsafe_function` have been stopped. That is, one would call `uv_thread_join()` fir all threads first, and then `napi_delete_threadsafe_function()`.

I guess I was also leaning on documentation whereby `napi_env` must only be used on the main thread to indicate that `napi_delete_threadsafe_function()` must only be called from the main thread.